internal/provider/resource_application.go

```diff
@@ -2,340 +2,536 @@ package provider
+	}
+	if val := n.Spaces.ValueString(); val != "" {
+		expose[SpacesKey] = val
+	}
+	if val := n.Cidrs.ValueString(); val != "" {
+		expose[CidrsKey] = val
+	}
+	return expose
+}
+
+func parseNestedExpose(value map[string]interface{}) nestedExpose {
+	// an empty expose structure, indicates exposure
+	// the values are optional.
+	resp := nestedExpose{}
+	if cidrs, ok := value[CidrsKey]; ok && cidrs != "" {
+		resp.Cidrs = types.StringValue(cidrs.(string))
+	}
+	if endpoints, ok := value[EndpointsKey]; ok && endpoints != "" {
+		resp.Endpoints = types.StringValue(endpoints.(string))
+	}
+	if spaces, ok := value[SpacesKey]; ok && spaces != "" {
+		resp.Spaces = types.StringValue(spaces.(string))
+	}
+	return resp
+}
+
+// Create is called when the provider must create a new resource. Config
+// and planned state values should be read from the
+// CreateRequest and new state values set on the CreateResponse.
+func (r *applicationResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	// Prevent panic if the provider has not been configured.
+	if r.client == nil {
+		addClientNotConfiguredError(&resp.Diagnostics, "application", "create")
+		return
+	}
+
+	var plan applicationResourceModel
+
+	// Read Terraform plan into the model
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	r.trace("Create", applicationResourceModelForLogging(ctx, &plan))
 
-	modelName := d.Get("model").(string)
-	modelUUID, err := client.Models.ResolveModelUUID(modelName)
+	modelInfo, err := r.client.Models.GetModelByName(plan.ModelName.ValueString())
 	if err != nil {
-		return diag.FromErr(err)
-	}
-
-	name := d.Get("name").(string)
-	charm := d.Get("charm").([]interface{})[0].(map[string]interface{})
-	charmName := charm["name"].(string)
-	channel := charm["channel"].(string)
-	if channel == "" {
-		// if no channel was set we will request the stable by default
-		channel = "stable"
-	}
-	series := charm["series"].(string)
-	units := d.Get("units").(int)
-	trust := d.Get("trust").(bool)
-	placement := d.Get("placement").(string)
-	// populate the config parameter
-	// terraform only permits a single type. We have to treat
-	// strings to have different types
-	configField := d.Get("config").(map[string]interface{})
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to get model info, got error: %s", err))
+		return
+	}
 
-	// if expose is nil, it was not defined
-	var expose map[string]interface{} = nil
-	exposeField, exposeWasSet := d.GetOk("expose")
-	if exposeWasSet {
-		// this was set, by default get no fields there
-		expose = make(map[string]interface{}, 0)
-		aux := exposeField.([]interface{})[0]
-		if aux != nil {
-			expose = aux.(map[string]interface{})
-		}
+	charms := []nestedCharm{}
+	resp.Diagnostics.Append(plan.Charm.ElementsAs(ctx, &charms, false)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	planCharm := charms[0]
```

Comment:
Charm is required and can be at most 1. Therefore it will exist.

` Block charm must have a configuration value as the provider has marked it as required`

Reply:
Ok, tried again, I think we're fine, I thought we could just make it an empty list:

```tf
charm {}
```

but looks like the "required" on the name field saves it:

``` The argument "name" is required, but no definition was found.```

Reply:
You may mark this as resolved when you see it